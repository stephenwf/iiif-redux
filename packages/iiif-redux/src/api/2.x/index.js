import annotationList, {
  mappings as annotationListMapping,
} from './annotation-list';
import annotation, { mappings as annotationMapping } from './annotation';
import canvas, { mappings as canvasMapping } from './canvas';
import collection, { mappings as collectionMapping } from './collection';
import externalResource, {
  mappings as externalResourceMapping,
} from './external-resource';
import imageResource, {
  mappings as imageResourceMapping,
} from './image-resource';
import manifest, { mappings as manifestMapping } from './manifest';
import range, { mappings as rangeMapping } from './range';
import sequence, { mappings as sequenceMapping } from './sequence';

export const mappings = {
  annotationList: annotationListMapping,
  annotation: annotationMapping,
  canvas: canvasMapping,
  collection: collectionMapping,
  externalResource: externalResourceMapping,
  imageResource: imageResourceMapping,
  manifest: manifestMapping,
  range: rangeMapping,
  sequence: sequenceMapping,

  // Aliases.
  annotationPage: annotationListMapping,
  contentResources: imageResourceMapping, // @todo change when image/external get merged.
};

export {
  annotationCollection,
  annotationList,
  annotation,
  canvas,
  collection,
  externalResource,
  imageResource,
  manifest,
  range,
  sequence,
};
