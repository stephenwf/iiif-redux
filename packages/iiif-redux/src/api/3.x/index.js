import annotationCollection, {
  mappings as annotationCollectionMapping,
} from './annotation-collection';
import annotationPage, {
  mappings as annotationPageMapping,
} from './annotation-page';
import annotation, { mappings as annotationMapping } from './annotation';
import canvas, { mappings as canvasMapping } from './canvas';
import collection, { mappings as collectionMapping } from './collection';
import contentResource, {
  mappings as contentResourceMapping,
} from './content-resource';
import manifest, { mappings as manifestMapping } from './manifest';
import range, { mappings as rangeMapping } from './range';

export const mappings = {
  annotationCollection: annotationCollectionMapping,
  annotationPage: annotationPageMapping,
  annotation: annotationMapping,
  canvas: canvasMapping,
  collection: collectionMapping,
  contentResource: contentResourceMapping,
  manifest: manifestMapping,
  range: rangeMapping,
};

export {
  annotationCollection,
  annotationPage,
  annotation,
  canvas,
  collection,
  contentResource,
  manifest,
  range,
};
