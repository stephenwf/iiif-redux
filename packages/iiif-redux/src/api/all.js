const getAllCollections = state => state.resources.collections;
const getAllExternalResources = state => state.resources.externalResources;
const getAllLayers = state => state.resources.layers;
const getAllServices = state => state.resources.services;
const getAllManifests = state => state.resources.manifests;
const getAllSequences = state => state.resources.sequences;
const getAllRanges = state => state.resources.ranges;
const getAllCanvases = state => state.resources.canvases;
const getAllCanvasReferences = state => state.resources.canvasReferences;
const getAllImages = state => state.resources.imageResources;
const getAllAnnotations = state => state.resources.annotations;
const getAllAnnotationLists = state => state.resources.annotationLists;
const getAllAnnotationPages = state => state.resources.annotationPages;
const getAllAnnotationCollections = state =>
  state.resources.annotationCollections;
const getAllContentResources = state => state.resources.contentResources;
const getAllResources = state => state.resources;

export {
  getAllCollections,
  getAllExternalResources,
  getAllLayers,
  getAllServices,
  getAllManifests,
  getAllSequences,
  getAllRanges,
  getAllCanvases,
  getAllCanvasReferences,
  getAllImages,
  getAllAnnotationLists,
  getAllAnnotationPages,
  getAllAnnotationCollections,
  getAllAnnotations,
  getAllContentResources,
  getAllResources,
};
