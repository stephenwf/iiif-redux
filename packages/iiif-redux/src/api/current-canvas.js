import { createSelector } from 'reselect';
import validUrl from 'valid-url';
import * as technical from './iiif-technical';
import * as descriptive from './iiif-descriptive';
import * as linking from './iiif-linking';
import * as structural from './iiif-structural';
import { getCurrentCanvas } from './current';
import {
  getAllAnnotationLists,
  getAllAnnotations,
  getAllExternalResources,
  getAllImages,
  getAllLayers,
  getAllServices,
} from './all';
import { isImageService } from '../constants/services';

/**************************************************
 * Technical properties
 *
 * - getId
 * - getType
 * - getViewingHint
 * - getHeight
 * - getWidth
 **************************************************/
const getId = createSelector(getCurrentCanvas, technical.getId);

const getType = createSelector(getCurrentCanvas, technical.getType);

const getViewingHint = createSelector(getCurrentCanvas, canvas => {
  const viewingHint = technical.getViewingHint(canvas);
  switch (viewingHint) {
    case 'non-paged':
    case 'facing-pages':
      return viewingHint;
    default:
      return validUrl.isWebUri(viewingHint) ? viewingHint : null;
  }
});

const getHeight = createSelector(getCurrentCanvas, technical.getHeight);

const getWidth = createSelector(getCurrentCanvas, technical.getWidth);

/**************************************************
 * Descriptive properties
 *
 * - getLabel
 * - getMetadata
 * - getDescription
 * - getThumbnail
 * - getAttribution
 * - getLicense
 * - getLogo
 **************************************************/
const getLabel = createSelector(getCurrentCanvas, descriptive.getLabel);

const getDescription = createSelector(
  getCurrentCanvas,
  descriptive.getDescription
);

const getMetadata = createSelector(getCurrentCanvas, descriptive.getMetadata);

const getAttribution = createSelector(
  getCurrentCanvas,
  descriptive.getAttribution
);

const getLogo = createSelector(getCurrentCanvas, descriptive.getLogo);

const getLicense = createSelector(getCurrentCanvas, descriptive.getLicense);

const getThumbnailId = createSelector(
  getCurrentCanvas,
  descriptive.getThumbnailId
);
const getThumbnail = createSelector(
  getThumbnailId,
  getAllImages,
  (thumbnailId, allImages) => allImages[thumbnailId] || thumbnailId
);

/**************************************************
 * Linking properties
 *
 * - getSeeAlso
 * - getService
 * - getRelated
 * - getRendering
 * - getWithin
 **************************************************/
const getSeeAlsoIds = createSelector(getCurrentCanvas, linking.getSeeAlso);
const getSeeAlso = createSelector(
  getSeeAlsoIds,
  getAllExternalResources,
  (seeAlsoIds, allExternalResources) =>
    seeAlsoIds.map(
      seeAlsoId =>
        allExternalResources[seeAlsoId] || {
          '@id': seeAlsoId,
          label: 'unknown',
        }
    )
);

const getServiceIds = createSelector(getCurrentCanvas, linking.getService);
const getService = createSelector(
  getServiceIds,
  getAllServices,
  (serviceIds, allServices) =>
    serviceIds.map(serviceId => allServices[serviceId])
);

const getRelatedIds = createSelector(getCurrentCanvas, linking.getRelated);
const getRelated = createSelector(
  getRelatedIds,
  getAllExternalResources,
  (relatedIds, allExternalResources) =>
    relatedIds.map(relatedId => allExternalResources[relatedId])
);

const getRenderingIds = createSelector(getCurrentCanvas, linking.getRendering);
const getRendering = createSelector(
  getRenderingIds,
  getAllExternalResources,
  (renderingIds, allExternalResources) =>
    renderingIds.map(renderingId => allExternalResources[renderingId])
);

const getWithinIds = createSelector(getCurrentCanvas, linking.getWithin);
const getWithin = createSelector(
  getWithinIds,
  getAllLayers,
  getAllExternalResources,
  (withinIds, allLayers, allExternalResources) =>
    withinIds.map(
      withinId => allLayers[withinId] || allExternalResources[withinId]
    )
);

/**************************************************
 * Structural properties
 *
 * - getOtherContent
 * - getImages
 **************************************************/
const getOtherContentIds = createSelector(
  getCurrentCanvas,
  structural.getOtherContent
);
const getOtherContent = createSelector(
  getOtherContentIds,
  getAllExternalResources,
  getAllAnnotationLists,
  (otherContentIds, allExternalResources, allAnnotationLists) =>
    otherContentIds.map(
      otherContentId =>
        allAnnotationLists[otherContentId] ||
        allExternalResources[otherContentId] || {
          '@id': otherContentId,
          label: 'unknown',
        }
    )
);

const getImageIds = createSelector(getCurrentCanvas, structural.getImages);
const getImages = createSelector(
  getImageIds,
  getAllAnnotations,
  (imageIds, allImages) => imageIds.map(imageId => allImages[imageId])
);

/**************************************************
 * Algorithms
 *
 * - getImageService
 **************************************************/
const getImageService = createSelector(
  getImages,
  getAllServices,
  getAllImages,
  (images, allServices, allImages) =>
    // The basic path is images[x].resource.service
    images.reduce((result, image) => {
      if (result) return result;
      const resource =
        image.resource.schema === 'imageResource'
          ? allImages[image.resource.id]
          : { service: [] };

      return resource.service.reduce((innerResult, serviceId) => {
        if (innerResult) return innerResult;
        const service = allServices[serviceId];
        if (isImageService(service.profile)) {
          return service;
        }
      }, null);
    }, null)
);

export {
  getCurrentCanvas,
  // Technical
  getId,
  getType,
  getViewingHint,
  getHeight,
  getWidth,
  // Descriptive
  getLabel,
  getMetadata,
  getDescription,
  getThumbnailId,
  getThumbnail,
  getAttribution,
  getLicense,
  getLogo,
  // Linking
  getWithinIds,
  getWithin,
  getRenderingIds,
  getRendering,
  getRelatedIds,
  getRelated,
  getServiceIds,
  getService,
  getSeeAlsoIds,
  getSeeAlso,
  // Structural
  getOtherContentIds,
  getOtherContent,
  getImageIds,
  getImages,
  // Algorithms.
  getImageService,
};
