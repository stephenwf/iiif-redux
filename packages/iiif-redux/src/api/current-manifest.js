import { getCurrentManifest } from './current';
import manifest from './manifest';

const {
  // Technical
  getId,
  getType,
  getViewingHint,
  getViewingDirection,
  getNavDate,
  // Descriptive
  getLabel,
  getDescription,
  getMetadata,
  getAttribution,
  getLogo,
  getLicense,
  getThumbnailId,
  getThumbnail,
  // Linking
  getWithinIds,
  getWithin,
  getRenderingIds,
  getRendering,
  getRelatedIds,
  getRelated,
  getServiceIds,
  getService,
  getSeeAlsoIds,
  getSeeAlso,
  // Structural
  getSequenceIds,
  getSequences,
  getRangeIds,
  getRanges,
  getStructureIds,
  getStructures,
  getOtherContentIds,
  getOtherContent,
} = manifest(getCurrentManifest);

export {
  getCurrentManifest,
  // Technical
  getId,
  getType,
  getViewingHint,
  getViewingDirection,
  getNavDate,
  // Descriptive
  getLabel,
  getDescription,
  getMetadata,
  getAttribution,
  getLogo,
  getLicense,
  getThumbnailId,
  getThumbnail,
  // Linking
  getWithinIds,
  getWithin,
  getRenderingIds,
  getRendering,
  getRelatedIds,
  getRelated,
  getServiceIds,
  getService,
  getSeeAlsoIds,
  getSeeAlso,
  // Structural
  getSequenceIds,
  getSequences,
  getRangeIds,
  getRanges,
  getStructureIds,
  getStructures,
  getOtherContentIds,
  getOtherContent,
};
