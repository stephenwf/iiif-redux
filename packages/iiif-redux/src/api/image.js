import memoize from 'lodash.memoize';
import { createSelector, createStructuredSelector } from 'reselect';
import * as technical from './iiif-technical';
import * as descriptive from './iiif-descriptive';
import * as structural from './iiif-structural';
import {
  getAllExternalResources,
  getAllImages,
  getAllLayers,
  getAllServices,
} from './all';
import * as linking from './iiif-linking';
import { isImageService } from '../constants/services';

const image = memoize(selector => {
  /**************************************************
   * Technical properties
   *
   * - getId
   * - getType
   * - getViewingHint
   **************************************************/
  const getId = createSelector(selector, technical.getId);

  const getType = createSelector(selector, technical.getType);

  const getViewingHint = createSelector(
    selector,
    technical.getWhitelistedViewingHint([])
  );

  /**************************************************
   * Descriptive properties
   *
   * - getLabel
   * - getMetadata
   * - getDescription
   * - getThumbnail
   * - getAttribution
   * - getLicense
   * - getLogo
   **************************************************/
  const getLabel = createSelector(selector, descriptive.getLabel);

  const getDescription = createSelector(selector, descriptive.getDescription);

  const getMetadata = createSelector(selector, descriptive.getMetadata);

  const getAttribution = createSelector(selector, descriptive.getAttribution);

  const getLogo = createSelector(selector, descriptive.getLogo);

  const getLicense = createSelector(selector, descriptive.getLicense);

  const getThumbnailId = createSelector(selector, descriptive.getThumbnailId);
  const getThumbnail = createSelector(
    getThumbnailId,
    getAllImages,
    (thumbnailId, allImages) => allImages[thumbnailId] || thumbnailId
  );

  /**************************************************
   * Linking properties
   *
   * - getSeeAlso
   * - getService
   * - getRelated
   * - getRendering
   * - getWithin
   **************************************************/
  const getSeeAlsoIds = createSelector(selector, linking.getSeeAlso);
  const getSeeAlso = createSelector(
    getSeeAlsoIds,
    getAllExternalResources,
    (seeAlsoIds, allExternalResources) =>
      seeAlsoIds.map(
        seeAlsoId =>
          allExternalResources[seeAlsoId] || {
            '@id': seeAlsoId,
            label: [{ '@value': 'unknown', '@language': '@none' }],
          }
      )
  );

  const getServiceIds = createSelector(selector, linking.getService);
  const getService = createSelector(
    getServiceIds,
    getAllServices,
    (serviceIds, allServices) =>
      serviceIds.map(serviceId => allServices[serviceId])
  );

  const getRelatedIds = createSelector(selector, linking.getRelated);
  const getRelated = createSelector(
    getRelatedIds,
    getAllExternalResources,
    (relatedIds, allExternalResources) =>
      relatedIds.map(relatedId => allExternalResources[relatedId])
  );

  const getRenderingIds = createSelector(selector, linking.getRendering);
  const getRendering = createSelector(
    getRenderingIds,
    getAllExternalResources,
    (renderingIds, allExternalResources) =>
      renderingIds.map(renderingId => allExternalResources[renderingId])
  );

  const getWithinIds = createSelector(selector, linking.getWithin);
  const getWithin = createSelector(
    getWithinIds,
    getAllLayers,
    getAllExternalResources,
    (withinIds, allLayers, allExternalResources) =>
      withinIds
        .map(within => allLayers[within.id] || allExternalResources[within.id])
        .filter(e => e)
  );

  /**************************************************
   * Structural properties
   *
   * - getResource
   **************************************************/
  const getResourceId = createSelector(selector, structural.getResource);
  const getResource = createSelector(
    getResourceId,
    getAllImages,
    (resourceId, allImageResources) => allImageResources[resourceId] || null
  );

  const getImageService = createSelector(
    getResource,
    getAllServices,
    (resource, allServices) =>
      resource
        ? resource.service.reduce((innerResult, serviceId) => {
            if (innerResult) return innerResult;
            const service = allServices[serviceId];
            if (isImageService(service.profile)) {
              return service;
            }
            return null;
          }, null)
        : null
  );

  return {
    // Technical
    getId,
    getType,
    getViewingHint,
    // Descriptive
    getLabel,
    getMetadata,
    getDescription,
    getThumbnailId,
    getThumbnail,
    getAttribution,
    getLicense,
    getLogo,
    // Linking
    getSeeAlso,
    getSeeAlsoIds,
    getService,
    getServiceIds,
    getRelated,
    getRelatedIds,
    getRendering,
    getRenderingIds,
    getWithin,
    getWithinIds,
    // Structural
    getResourceId,
    getResource,
    // Service
    getImageService,
  };
});

export default image;

export function imageByIdSelector(callable, getId) {
  return (state, props) =>
    createStructuredSelector(
      callable(
        image(
          () => state.resources.annotations[getId ? getId(props) : props.id]
        )
      )
    )(state);
}
