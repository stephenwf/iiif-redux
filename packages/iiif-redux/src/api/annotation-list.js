import { getAllAnnotationLists } from './all';
import byIdSelectorFactory from '../utility/byIdSelectorFactory';
import resourceListSelectorFactory from '../utility/resourceListSelectorFactory';
import presentation2annotationList from './2.x/annotation-list';

const annotationList = presentation2annotationList;

export default annotationList;

export const annotationLists = resourceListSelectorFactory(
  getAllAnnotationLists,
  annotationList
);

export const annotationListByIdSelector = byIdSelectorFactory(
  annotationList,
  'annotationLists'
);
