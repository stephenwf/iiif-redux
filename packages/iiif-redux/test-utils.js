import createStore from './src/createStore';
import {
  selectCanvas,
  selectManifest,
  selectSequence,
} from './src/spaces/routing';
import * as currentManifest from './src/api/current-manifest';
import * as currentSequence from './src/api/current-sequence';
import * as currentCanvas from './src/api/current-canvas';

export function waitForRequest(store, id) {
  return new Promise(resolve => {
    store.subscribe(() => {
      const newState = store.getState();
      if (
        newState.dereferenced &&
        newState.dereferenced[id] &&
        newState.dereferenced[id].loading === false
      ) {
        resolve();
      }
    });
  });
}

export async function createStoreAndImportManifest(manifestJson) {
  const id = manifestJson['@id'];
  const store = createStore();
  fetch.mockResponseOnce(JSON.stringify(manifestJson));

  const whenRequestFinishes = waitForRequest(store, id);

  store.dispatch(selectManifest({ id }));

  await whenRequestFinishes;

  const sequences = currentManifest.getSequenceIds(store.getState());
  store.dispatch(selectSequence({ id: sequences[0] }));

  const canvases = currentSequence.getCanvasIds(store.getState());
  store.dispatch(selectCanvas({ id: canvases[0] }));

  return store;
}
