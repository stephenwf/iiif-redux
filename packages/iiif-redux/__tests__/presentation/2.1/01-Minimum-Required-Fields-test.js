import { createStructuredSelector } from 'reselect';
import testJson from '../../fixtures/presentation/2.1/01/manifest';
import * as currentManifest from '../../../src/api/current-manifest';
import * as currentCanvas from '../../../src/api/current-canvas';
import { imageByIdSelector } from '../../../src/api/image';
import { createStoreAndImportManifest } from '../../../test-utils';

describe('iiif/presentation/2.1/01 Minimum Required Fields', () => {
  global.fetch = require('jest-fetch-mock');

  it('should render all fields on manifest', async () => {
    const store = await createStoreAndImportManifest(testJson);
    const state = store.getState();

    expect(
      createStructuredSelector({
        manifest: createStructuredSelector({
          id: currentManifest.getId,
          type: currentManifest.getType,
          label: currentManifest.getLabel,
          within: currentManifest.getWithin,
          sequences: currentManifest.getSequences,
        }),
        canvas: createStructuredSelector({
          // Canvas
          id: currentCanvas.getId,
          label: currentCanvas.getLabel,
          height: currentCanvas.getHeight,
          width: currentCanvas.getWidth,
          images: currentCanvas.getImages,
          imageService: currentCanvas.getImageService,
          image: imageByIdSelector(
            image => ({
              id: image.getId,
              motivation: image.getMotivation,
              on: image.getOn,
              resource: image.getResource,
            }),
            () => 'https://23b7cac9e952d1868e598100cd25352f482db919'
          ),
        }),
      })(state)
    ).toMatchSnapshot();
  });
});
