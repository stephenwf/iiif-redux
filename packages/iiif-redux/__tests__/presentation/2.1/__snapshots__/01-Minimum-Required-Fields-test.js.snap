// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`iiif/presentation/2.1/01 Minimum Required Fields should render all fields on manifest 1`] = `
Object {
  "canvas": Object {
    "height": 1800,
    "id": "http://iiif.io/api/presentation/2.1/example/fixtures/canvas/1/c1.json",
    "image": Object {
      "id": "https://23b7cac9e952d1868e598100cd25352f482db919",
      "motivation": "sc:painting",
      "on": Object {
        "@id": "http://iiif.io/api/presentation/2.1/example/fixtures/canvas/1/c1.json",
        "@type": "sc:Canvas",
        "height": 1800,
        "images": Array [
          "https://23b7cac9e952d1868e598100cd25352f482db919",
        ],
        "label": Array [
          Object {
            "@language": "en",
            "@value": "Test 1 Canvas: 1",
          },
        ],
        "width": 1200,
      },
      "resource": Object {
        "@id": "http://iiif.io/api/presentation/2.1/example/fixtures/resources/page1-full.png",
        "@type": "dctypes:Image",
        "height": 1800,
        "width": 1200,
      },
    },
    "imageService": null,
    "images": Array [
      Object {
        "@id": "https://23b7cac9e952d1868e598100cd25352f482db919",
        "@type": "oa:Annotation",
        "motivation": "sc:painting",
        "on": "http://iiif.io/api/presentation/2.1/example/fixtures/canvas/1/c1.json",
        "resource": Object {
          "id": "http://iiif.io/api/presentation/2.1/example/fixtures/resources/page1-full.png",
          "schema": "imageResource",
        },
      },
    ],
    "label": Array [
      Object {
        "@language": "en",
        "@value": "Test 1 Canvas: 1",
      },
    ],
    "width": 1200,
  },
  "manifest": Object {
    "id": "http://iiif.io/api/presentation/2.1/example/fixtures/1/manifest.json",
    "label": Array [
      Object {
        "@language": "en",
        "@value": "Test 1 Manifest: Minimum Required Fields",
      },
    ],
    "sequences": Array [
      Object {
        "@id": "https://1860e2c42d700c69bb3bf0291656fe1b7689e6f2",
        "@type": "sc:Sequence",
        "canvases": Array [
          "http://iiif.io/api/presentation/2.1/example/fixtures/canvas/1/c1.json",
        ],
      },
    ],
    "type": "sc:Manifest",
    "within": Array [
      Object {
        "@id": "http://iiif.io/api/presentation/2.1/example/fixtures/collection.json",
      },
    ],
  },
}
`;
